Switch all tools to TPV

kysrpex opened the pull request:
Closes [#383 from the infrastructure issues repository](https://github.com/usegalaxy-eu/issues/issues/383).

- [x] Add new changes in dynamic destinations to the TPV shared database.
- [x] Migrate marked destinations from sorting hat to TPV.
- [x] Add scheduling tags for affected tools.
- [x] Switch default dispatcher.

kysrpex commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/tools.yml:
```diff
@@ -39,6 +39,14 @@ tools:
       TEMP: /data/1/galaxy_db/tmp
 
+  gmx_sim:
+    gpus: 1
```
@mira-miracoli @sanjaysrikakulam @cat-bro Is the setting `resources: usegpu` in job_conf.xml equivalent to `gpus: 1` in tools.yml?

kysrpex replied:
@mira-miracoli @sanjaysrikakulam According to @cat-bro, `gpus: 1` in TPV means "the destination has at least 1 gpu", so you would just need to confirm whether `resources: usegpu` has exactly the same meaning.

sanjaysrikakulam replied:
IIRC: the following should be added 

```
scheduling:
      require:
        - condor-gpu
```

mira-miracoli replied:
yes we discussed this, because our experience was that gpu destinations also take jobs that don't specify GPUs, which we would like to avoid. (or did I get something wrong @cat-bro)

kysrpex replied:
I set up the rule-based solution in [6808149](https://github.com/usegalaxy-eu/infrastructure-playbook/pull/689/commits/68081490337205a94ea641a9b1fcba1f3366d27a), closing the conversation.

kysrpex commented:
@mira-miracoli I am pointing the tip of the branch to [d79aeec](https://github.com/usegalaxy-eu/infrastructure-playbook/pull/689/commits/d79aeecef1e12f89f6c4b6d7b90e24313b9630f9) again.

Is there any benefit that I may have missed in merging the changes like this? I can always point to [b97274c](https://github.com/usegalaxy-eu/infrastructure-playbook/pull/689/commits/b97274c2322a60cb6f51622031311e501444fb3d) again if it is the case.

mira-miracoli commented:
sorry my fault should do a merge next time

bgruening commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/tools.yml:
```diff
@@ -25,6 +25,11 @@ tools:
+        if: gpus > 0
+        scheduling:
+          require:
+            - condor-gpu
```
(I'm not sure this fits into this PR)
I think we should think carefully about the ontology of tags that we are using here and discuss this with @cat-bro . Having something like condor in the tag name might not be portable to instances (others than EU) that do not have condor. I think I would like to see tags very generic in their ontology so that we can share as much as possible (tool def, destinations ...) with other instances.

bgruening replied:
Sure, it probably also needs some discussions and an agreed terminology.

kysrpex replied:
We actually did some changes regarding this that are still on the VSCode server, they should show up once we commit everything.

kysrpex commented:
Tomorrow I will run `tpv dry-run` and post the results to make sure things work as expected (at least in theory). If that goes well then I will remove the draft status to ask for review I guess? I hope then you can help me out pointing out if I missed something.

bgruening commented:
Isn't the dry-run running on this PR? If not it should I think.

kysrpex commented:
> Isn't the dry-run running on this PR? If not it should I think.

Just [yamllint](https://github.com/kysrpex/usegalaxy-eu-ansible/blob/0c538bfcb4236f0914eeaad0ccc14571d451d06a/.github/workflows/ci.yml#L33), but I will add it, it makes a lot of sense.

kysrpex commented on templates/galaxy/config/job_conf.yml:
```diff
@@ -501,507 +465,9 @@ galaxy_jobconf:
-      destination: tpv_dispatcher
-    - id: gops_coverage_1
-      destination: tpv_dispatcher
   resources:
     - id: usegpu
       names: [gpu]
```
Is this even needed anymore?

kysrpex replied:
Thanks @cat-bro for clarifying that it is no longer needed.

mira-miracoli commented:
Can the old `condor_docker_gpu` destination be fully replaced by interactive pulsar GPU or did I miss something?

mira-miracoli commented:
maybe a list with tools that have different tags, if the tags and gpus are the same maybe we can ignore other differences

bgruening commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,105 +1,76 @@
 ---
-# ALL tags must be with dashes (-) instead of underscores (_)
+# NOTE: Use dashes (-) exclusively for tags and underscores (_) exclusively for destinations.
```
This note is great. I would also move such rules in the readme of shared-tpv repo.

kysrpex replied:
Well destinations are not a thing on the shared-tpv repo, only scheduling tags, and the [README already says](https://github.com/galaxyproject/tpv-shared-database/blob/2c2e8596ad35c74b961a03df0d00e37a0b27e841/README.md?plain=1#L15).

> Dashes are used as separators in tags instead of underscores.

So I guess this is done already.

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,97 +1,76 @@
+  # EMBEDDED PULSAR DESTINATIONS #
+  ################################
+
+  interactive_pulsar:
```
```suggestion
  embedded_pulsar:
```
could we use embedded more generic and not only for interactive tools, or is there a reason to limit it?

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,97 +1,76 @@
-      prefer:
+      accept:
         - docker
+      require:
```
```suggestion
      prefer:
```

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,97 +1,76 @@
         - docker
+      require:
         - interactive
 
```
```suggestion
        - embedded-pulsar
  
```

bgruening commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,105 +1,76 @@
 destinations:
-  interactive_pulsar:
-    runner: pulsar_embedded
-    destination_name_override: "embedded_pulsar_interactive_c{cores}-m{mem}"
```
Maybe we can think about a unified naming schema? The above is too long, but I could imagine something like:

## destination

* ep = embedded pulsar
* p = pulsar
* c = condor
* s = slurm
* k = k8s

## resolver

* d = docker
* s = singularity
* c = conda

## resources

* cXX = cores
* mYY = memory
* gZZ = GPUs
* n = network (e.g. interactive tools)

could lead to something like `{destination}_{resolver}_{resources}`

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,97 +1,76 @@
     scheduling:
       prefer:
         - docker
         - embedded-pulsar
```
this could be removed in that case

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,97 +1,76 @@
+      require:
         - interactive
 
   interactive_pulsar_gpu:
```
```suggestion
  embedded_pulsar_gpu:
```

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -109,122 +88,125 @@ destinations:
+    params:
+      metadata_strategy: "extended"
+    scheduling:
+      prefer:
```
```suggestion
      require:
```
just to make sure?

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -109,122 +88,125 @@ destinations:
-    #  GALAXY_SLOTS: '{PARALLELISATION}'
     params:
-      priority: '{PRIORITY}'
-      request_cpus: '{cores}'
```
Inherited from default tool

bgruening commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,105 +1,76 @@
+      submit_request_memory: "{mem}G"
+      singularity_enabled: true
+      singularity_volumes: "$_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dnb08/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:r ...
+      singularity_default_container_id: "/cvmfs/singularity.galaxyproject.org/all/centos:8.3.2011"
```
we should agree on one standard image in the long run. Should this be centos, ubuntu, which version?

bgruening commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,105 +1,76 @@
-    scheduling:
-      require:
-        - extended-metadata
+  #######################
```
would it help or only confuse if we split those up into separate files?

mira-miracoli replied:
I would try to stick to few files

bgruening commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -117,122 +88,127 @@ destinations:
+    scheduling:
+      require:
+        - singularity
+        - conda
```
Does that mean that all tools needs to have a tag=conda?

mira-miracoli replied:
yes because it is a destination that is only used by tools that want to run conda inside singularity

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/tools.yml:
```diff
@@ -606,3 +669,181 @@ tools:
+        - singularity
+        - conda
+  # is there a way to avoid this
+  "*.pcgr.*":
```
drop? I did not find a long name.

bgruening replied:
It is this tool here, that we deploy locally https://github.com/bgruening/galaxytools/pull/1257  
@wm75 do we have plans for it?

kysrpex replied:
If no long id neither exists nor is planned by @wm75, then the following should work.

```suggestion
  ^pcgr$:
```

mira-miracoli commented:
> > Isn't the dry-run running on this PR? If not it should I think.
> 
> I included the linting (although I should solve an issue yet to make it faster). However, does the dry-run make any sense to be included in the CI? It has to be run for each tool and it is infrastructure-dependent, since checking that it works properly involves checking that the tool is sent to the expected destination.
> 
> Maybe running the [playbook from @cat-bro](https://github.com/cat-bro/infrastructure-playbook/pull/2) a couple of times for a few tools and checking that they go to the appropriate destination makes better sense?
My suggestion for a tool list:
```yaml
  # has a GPU but nothing else
  - toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim
  #  a legacy tool that wants either containers nor embedded pulsar
  - toolshed.g2.bx.psu.edu/repos/iuc/snippy/snippy
  # a tool that wants to run conda in singularity
  - toolshed.g2.bx.psu.edu/repos/iuc/enasearch_search_data/enasearch_search_data
  # singularity only
  - toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold
  # a gpu from shared DB and docker
  - toolshed.g2.bx.psu.edu/repos/bgruening/instagraal/instagraal
  # should go to condor-tpv
  - toolshed.g2.bx.psu.edu/repos/nml/metaspades/metaspades
  # embedded and docker
  - mothur_*
  # docker but nothing else
  - toolshed.g2.bx.psu.edu/repos/bgruening/xchem_transfs_scoring/xchem_transfs_scoring
  # no tags
  - toolshed.g2.bx.psu.edu/repos/computational-metabolomics/dimspy_process_scans/dimspy_process_scans
```

kysrpex commented:
> > > Isn't the dry-run running on this PR? If not it should I think.
> > 
> > 
> > I included the linting (although I should solve an issue yet to make it faster). However, does the dry-run make any sense to be included in the CI? It has to be run for each tool and it is infrastructure-dependent, since checking that it works properly involves checking that the tool is sent to the expected destination.
> > Maybe running the [playbook from @cat-bro](https://github.com/cat-bro/infrastructure-playbook/pull/2) a couple of times for a few tools and checking that they go to the appropriate destination makes better sense?
> 
> My suggestion for a tool list:
> 
> ```yaml
>   # has a GPU but nothing else
>   - toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim
>   #  a legacy tool that wants either containers nor embedded pulsar
>   - toolshed.g2.bx.psu.edu/repos/iuc/snippy/snippy
>   # a tool that wants to run conda in singularity
>   - toolshed.g2.bx.psu.edu/repos/iuc/enasearch_search_data/enasearch_search_data
>   # singularity only
>   - toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold
>   # a gpu from shared DB and docker
>   - toolshed.g2.bx.psu.edu/repos/bgruening/instagraal/instagraal
>   # should go to condor-tpv
>   - toolshed.g2.bx.psu.edu/repos/nml/metaspades/metaspades
>   # embedded and docker
>   - mothur_*
>   # docker but nothing else
>   - toolshed.g2.bx.psu.edu/repos/bgruening/xchem_transfs_scoring/xchem_transfs_scoring
>   # no tags
>   - toolshed.g2.bx.psu.edu/repos/computational-metabolomics/dimspy_process_scans/dimspy_process_scans
> ```

Thanks for compiling the list! :+1:

kysrpex commented on group_vars/sn06.yml:
```diff
@@ -470,7 +470,7 @@ galaxy_dynamic_job_rules_src_dir: files/galaxy/dynamic_rules
 galaxy_dynamic_job_rules:
   - usegalaxy/joint_destinations.yaml
-  - usegalaxy/sorting_hat.py
+  - usegalaxy/sorting_hat.yaml
```
```suggestion
  - usegalaxy/sorting_hat.py
```

`usegalaxy/sorting_hat.yaml` does not exist...

cat-bro commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,97 +1,87 @@
+      submit_requirements: "GalaxyDockerHack == True"
+
+  basic_singularity_destination:
+    params:
```
Is this a destination that jobs can go to, or is it a base to be inherited by other destinations? If this is not a destination that jobs will go to, it can have `abstract: true` so that TPV will never consider it as a candidate destination.

mira-miracoli commented on files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml:
```diff
@@ -1,97 +1,90 @@
+    inherits: basic_singularity_destination
+    abstract: true
     runner: pulsar_embedded
     destination_name_override: "embedded_pulsar_c{cores}-m{mem}"
```
```suggestion
```

kysrpex commented:
> > > > Isn't the dry-run running on this PR? If not it should I think.
> > > 
> > > 
> > > I included the linting (although I should solve an issue yet to make it faster). However, does the dry-run make any sense to be included in the CI? It has to be run for each tool and it is infrastructure-dependent, since checking that it works properly involves checking that the tool is sent to the expected destination.
> > > Maybe running the [playbook from @cat-bro](https://github.com/cat-bro/infrastructure-playbook/pull/2) a couple of times for a few tools and checking that they go to the appropriate destination makes better sense?
> > 
> > 
> > My suggestion for a tool list:
> > ```yaml
> >   # has a GPU but nothing else
> >   - toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim
> >   #  a legacy tool that wants either containers nor embedded pulsar
> >   - toolshed.g2.bx.psu.edu/repos/iuc/snippy/snippy
> >   # a tool that wants to run conda in singularity
> >   - toolshed.g2.bx.psu.edu/repos/iuc/enasearch_search_data/enasearch_search_data
> >   # singularity only
> >   - toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold
> >   # a gpu from shared DB and docker
> >   - toolshed.g2.bx.psu.edu/repos/bgruening/instagraal/instagraal
> >   # should go to condor-tpv
> >   - toolshed.g2.bx.psu.edu/repos/nml/metaspades/metaspades
> >   # embedded and docker
> >   - mothur_*
> >   # docker but nothing else
> >   - toolshed.g2.bx.psu.edu/repos/bgruening/xchem_transfs_scoring/xchem_transfs_scoring
> >   # no tags
> >   - toolshed.g2.bx.psu.edu/repos/computational-metabolomics/dimspy_process_scans/dimspy_process_scans
> > ```
> 
> Thanks for compiling the list! +1

Would you say this mapping is correct?

```bash
TOOLS=(
        'toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim'  # has a GPU but nothing else
        'toolshed.g2.bx.psu.edu/repos/iuc/snippy/snippy'  # a legacy tool that wants neither containers nor embedded pulsar
        'toolshed.g2.bx.psu.edu/repos/iuc/enasearch_search_data/enasearch_search_data'  # a tool that wants to run conda in singularity
        'toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold'  # singularity only
        'toolshed.g2.bx.psu.edu/repos/bgruening/instagraal/instagraal'  # a gpu from shared DB and docker
        'toolshed.g2.bx.psu.edu/repos/nml/metaspades/metaspades'  # should go to condor-tpv
        'mothur_*'  # embedded and docker
        'toolshed.g2.bx.psu.edu/repos/bgruening/xchem_transfs_scoring/xchem_transfs_scoring' # docker but nothing else
        'toolshed.g2.bx.psu.edu/repos/computational-metabolomics/dimspy_process_scans/dimspy_process_scans' # no tags
)
DESTINATIONS=(
        (condor_gpu embedded_pulsar_docker_gpu)
        (condor_tpv condor_singularity)
        (condor_singularity_with_conda)
        (condor_singularity)
        (embedded_pulsar_docker_gpu)
        (condor_tpv)
        (embedded_pulsar_docker)
        (embedded_pulsar_docker)
        (embedded_pulsar_docker condor_tpv condor_singularity)
)
```

kysrpex commented:
Results of local dry run after the latest fixes:

```shell
$ for tool in "${TOOLS[@]}"; do echo "-----------------------------------------------"; echo "$tool/.*"; tpv dry-run --job-conf /home/jose/Dokumente/Repositories/usegalaxy-eu/infrastructure-playbook/job_conf.yml --tool "$tool/.*"; done
-----------------------------------------------
toolshed.g2.bx.psu.edu/repos/chemteam/gmx_sim/gmx_sim/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env:
- {name: GPU_AVAILABLE, value: '1'}
id: embedded_pulsar_docker_gpu
legacy: false
params: {container_monitor_result: callback, docker_auto_rm: true, docker_enabled: true,
  docker_memory: 30.4G, docker_net: bridge, docker_set_user: '', docker_sudo: false,
  metadata_strategy: extended, outputs_to_working_directory: false, request_cpus: '8',
  request_memory: 30.4G, require_container: true, requirements: GalaxyGroup == "compute_gpu",
  submit_request_cpus: '8', submit_request_gpus: '1', submit_request_memory: 30.4G,
  submit_requirements: GalaxyDockerHack == True, tmp_dir: true}
resubmit: []
runner: pulsar_embedded
shell: null
tags: null
url: null
-----------------------------------------------
toolshed.g2.bx.psu.edu/repos/iuc/snippy/snippy/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env: []
id: condor_tpv
legacy: false
params: {docker_memory: 12G, request_cpus: '2', request_memory: 12G, submit_request_gpus: '0',
  tmp_dir: true}
resubmit: []
runner: condor
shell: null
tags: null
url: null
-----------------------------------------------
toolshed.g2.bx.psu.edu/repos/iuc/enasearch_search_data/enasearch_search_data/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env: []
id: condor_singularity_with_conda
legacy: false
params:
  container_override:
  - {identifier: /data/0/singularity_base_images/centos:8.3.2011, resolve_dependencies: true,
    shell: /bin/bash, type: singularity}
  docker_memory: 3.8G
  request_cpus: '1'
  request_memory: 3.8G
  singularity_default_container_id: /cvmfs/singularity.galaxyproject.org/all/centos:8.3.2011
  singularity_enabled: true
  singularity_volumes: $_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dnb08/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:ro,/data/2/galaxy_db/:ro,/data/3/galaxy_db/:ro,/data/4/galaxy_db/:ro,/data/5/galaxy_import/galaxy_user_data/:ro,/data/6/galaxy_db/:ro,/data/7/galaxy_db/:ro,/usr/local/tools/:ro
  submit_request_cpus: '1'
  submit_request_gpus: '0'
  submit_request_memory: 3.8G
  tmp_dir: true
resubmit: []
runner: condor
shell: null
tags: null
url: null
-----------------------------------------------
toolshed.g2.bx.psu.edu/repos/galaxy-australia/alphafold2/alphafold/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env: []
id: condor_singularity
legacy: false
params: {docker_memory: 32G, metadata_strategy: extended, request_cpus: '10', request_memory: 32G,
  singularity_default_container_id: /cvmfs/singularity.galaxyproject.org/all/centos:8.3.2011,
  singularity_enabled: true, singularity_run_extra_arguments: '--env ALPHAFOLD_DB=/data/db/databases/alphafold_databases/2.2/,ALPHAFOLD_USE_GPU=False',
  singularity_volumes: '$_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dnb08/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:ro,/data/2/galaxy_db/:ro,/data/3/galaxy_db/:ro,/data/4/galaxy_db/:ro,/data/5/galaxy_import/galaxy_user_data/:ro,/data/6/galaxy_db/:ro,/data/7/galaxy_db/:ro,/usr/local/tools/:ro',
  submit_request_cpus: '10', submit_request_gpus: '0', submit_request_memory: 32G,
  tmp_dir: true}
resubmit: []
runner: condor
shell: null
tags: null
url: null
-----------------------------------------------
toolshed.g2.bx.psu.edu/repos/bgruening/instagraal/instagraal/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env:
- {name: GPU_AVAILABLE, value: '1'}
id: embedded_pulsar_docker_gpu
legacy: false
params: {container_monitor_result: callback, docker_auto_rm: true, docker_enabled: true,
  docker_memory: 30G, docker_net: bridge, docker_set_user: '', docker_sudo: false,
  metadata_strategy: extended, outputs_to_working_directory: false, request_cpus: '1',
  request_memory: 30G, require_container: true, requirements: GalaxyGroup == "compute_gpu",
  submit_request_cpus: '1', submit_request_gpus: '1', submit_request_memory: 30G,
  submit_requirements: GalaxyDockerHack == True, tmp_dir: true}
resubmit: []
runner: pulsar_embedded
shell: null
tags: null
url: null
-----------------------------------------------
toolshed.g2.bx.psu.edu/repos/nml/metaspades/metaspades/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env: []
id: condor_tpv
legacy: false
params: {docker_memory: 250G, request_cpus: '2', request_memory: 250G, submit_request_gpus: '0',
  tmp_dir: true}
resubmit: []
runner: condor
shell: null
tags: null
url: null
-----------------------------------------------
mothur_*/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env: []
id: embedded_pulsar_docker
legacy: false
params: {container_monitor_result: callback, docker_auto_rm: true, docker_default_container_id: centos:8.3.2011,
  docker_enabled: true, docker_memory: 90G, docker_net: bridge, docker_run_extra_arguments: --pids-limit
    10000 --ulimit fsize=1000000000 --env TERM=vt100, docker_set_user: '', docker_sudo: false,
  docker_volumes: '$_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:ro,/data/2/galaxy_db/:ro,/data/3/galaxy_db/:ro,/data/4/galaxy_db/:ro,/data/5/galaxy_import/galaxy_user_data/:ro,/data/6/galaxy_db/:ro,/data/7/galaxy_db/:ro,/usr/local/tools/:ro',
  object_store_id: files13, outputs_to_working_directory: false, request_cpus: '1',
  request_memory: 90G, require_container: true, submit_request_cpus: '1', submit_request_gpus: '0',
  submit_request_memory: 90G, submit_requirements: GalaxyDockerHack == True, tmp_dir: true}
resubmit: []
runner: pulsar_embedded
shell: null
tags: null
url: null
-----------------------------------------------
toolshed.g2.bx.psu.edu/repos/bgruening/xchem_transfs_scoring/xchem_transfs_scoring/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env:
- {name: CUDA_VISIBLE_DEVICES, value: '0'}
- {name: GPU_AVAILABLE, value: '1'}
id: embedded_pulsar_docker_gpu
legacy: false
params: {container_monitor_result: callback, docker_auto_rm: true, docker_enabled: true,
  docker_memory: 3.8G, docker_net: bridge, docker_set_user: '', docker_sudo: false,
  metadata_strategy: extended, outputs_to_working_directory: false, request_cpus: '1',
  request_memory: 3.8G, require_container: true, requirements: GalaxyGroup == "compute_gpu",
  submit_request_cpus: '1', submit_request_gpus: '1', submit_request_memory: 3.8G,
  submit_requirements: GalaxyDockerHack == True, tmp_dir: true}
resubmit: []
runner: pulsar_embedded
shell: null
tags: null
url: null
-----------------------------------------------
toolshed.g2.bx.psu.edu/repos/computational-metabolomics/dimspy_process_scans/dimspy_process_scans/.*
!!python/object:galaxy.jobs.JobDestination
converted: false
env:
- {name: NUMBA_CACHE_DIR, value: /data/2/galaxy_db/tmp}
- {name: OMP_NUM_THREADS, value: '4'}
- {name: OPENBLAS_NUM_THREADS, value: '4'}
- {name: MKL_NUM_THREADS, value: '4'}
- {name: VECLIB_MAXIMUM_THREADS, value: '4'}
- {name: NUMEXPR_NUM_THREADS, value: '4'}
- {name: NUMBA_NUM_THREADS, value: '4'}
id: condor_tpv
legacy: false
params: {docker_memory: 12G, request_cpus: '4', request_memory: 12G, submit_request_gpus: '0',
  tmp_dir: true}
resubmit: []
runner: condor
shell: null
tags: null
url: null 
```

**job_conf.yml** used for the local dry run
```yaml
---
##
## This file is maintained by Ansible - CHANGES WILL BE OVERWRITTEN
## Do not edit this file manually.
## Any changes will be automatically reverted.
##

#
# Job runner plugin configuration
#
runners:
  condor:
    load: galaxy.jobs.runners.condor:CondorJobRunner
  dynamic:
    rules_module: usegalaxy
  local:
    workers: 8
    load: galaxy.jobs.runners.local:LocalJobRunner
  pulsar_embedded:
    load: galaxy.jobs.runners.pulsar:PulsarEmbeddedJobRunner
    pulsar_config: "/opt/galaxy/config/pulsar_app.yml"
  pulsar_eu_be01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_be01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_be01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_cz01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_cz01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_cz01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_es01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_es01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_es01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_fi01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_fi01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_fi01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_fr01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_fr01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_fr01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_it01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_it01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_it01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_it02:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_it02:password@mq.galaxyproject.eu:5671//pulsar/galaxy_it02?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_it03:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_it03:password@mq.galaxyproject.eu:5671//pulsar/galaxy_it03?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_no01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_no01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_no01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_pt01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_pt01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_pt01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_sk01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_sk01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_sk01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_eu_uk01:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_uk01:password@mq.galaxyproject.eu:5671//pulsar/galaxy_uk01?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_mira_runner:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_mira_pulsar:password@mq.galaxyproject.eu:5671//pulsar/galaxy_mira_pulsar?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production
  pulsar_sanjay_runner:
    load: galaxy.jobs.runners.pulsar:PulsarMQJobRunner
    amqp_ack_republish_time: 300
    amqp_acknowledge: true
    amqp_consumer_timeout: 2.0
    amqp_publish_retry: true
    amqp_publish_retry_max_retries: 60
    amqp_url: pyamqp://galaxy_sanjay_pulsar:password@mq.galaxyproject.eu:5671//pulsar/galaxy_sanjay_pulsar?ssl=1
    galaxy_url: https://usegalaxy.eu
    manager: production

#
# Job handler configuration
#
handling:
  assign_with:
    - db-skip-locked
  max_grab: 16
  ready_window_size: 32
  processes:
    dummy_0:
    dummy_1:
    dummy_2:
    dummy_3:
    dummy_4:
    dummy_5:

#
# Job execution configuration
#
execution:
  default: tpv_dispatcher
  environments:
    condor:
      runner: condor
      tmp_dir: true
    condor_docker:
      runner: condor
      docker_auto_rm: true
      docker_default_container_id: ubuntu:18.04
      docker_enabled: true
      docker_set_user: 
      docker_sudo: false
      docker_volumes: $galaxy_root:ro,$tool_directory:ro,$working_directory:rw,$job_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro
    condor_docker_ie:
      runner: condor
      docker_auto_rm: true
      docker_default_container_id: ubuntu:18.04
      docker_enabled: true
      docker_set_user: 
      docker_sudo: false
      docker_volumes: $galaxy_root:ro,$tool_directory:ro,$working_directory:rw,$job_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro
    condor_docker_ie_interactive:
      runner: condor
      docker_auto_rm: true
      docker_default_container_id: ubuntu:18.04
      docker_enabled: true
      docker_set_user: 
      docker_sudo: false
      docker_volumes: $working_directory:rw,$job_directory:rw
    condor_docker_interactive:
      runner: condor
      docker_auto_rm: true
      docker_default_container_id: ubuntu:18.04
      docker_enabled: true
      docker_set_user: 
      docker_sudo: false
      docker_volumes: $working_directory:rw,$job_directory:rw
    condor_singularity:
      runner: condor
      singularity_default_container_id: /data/0/cvmfs/singularity/ubuntu:18.04
      singularity_enabled: true
      singularity_volumes: $_CONDOR_SCRATCH_DIR:rw,$job_directory:rw,$tool_directory:ro,$job_directory/outputs:rw,$working_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro,/data/dnb03/galaxy_db/:ro,/data/dnb05/galaxy_db/:ro,/data/dnb06/galaxy_db/:rw,/data/dnb07/galaxy_db/:rw,/data/dp01/galaxy_db/:rw,/data/0/galaxy_db/:ro,/data/1/galaxy_db/:ro,/data/2/galaxy_db/:ro,/data/3/galaxy_db/:ro,/data/4/galaxy_db/:ro,/data/5/galaxy_import/galaxy_user_data/:ro,/data/6/galaxy_db/:ro,/data/7/galaxy_db/:ro,/usr/local/tools/:ro
    condor_singularity2:
      runner: condor
      container: [{'type': 'singularity', 'shell': '/bin/sh', 'resolve_dependencies': 'true', 'identifier': '/data/0/cvmfs/singularity/ubuntu:18.04'}]
      singularity_default_container_id: /data/0/cvmfs/singularity/ubuntu:18.04
      singularity_enabled: true
      singularity_volumes: $galaxy_root:ro,$tool_directory:ro,$working_directory:rw,$job_directory:rw,/data/db/:ro,/data/dnb01/galaxy_db/:ro,/data/dnb02/galaxy_db/:ro
    condor_singularity_with_conda:
      runner: condor
      container: [{'type': 'singularity', 'shell': '/bin/bash', 'resolve_dependencies': 'true', 'identifier': '/data/0/cvmfs/singularity/python:2.7.12--1'}]
      singularity_enabled: true
    condor_tpv:
      runner: condor
      tmp_dir: true
    embedded_pulsar_docker:
      runner: pulsar_embedded
      container_monitor_result: callback
      docker_auto_rm: True
      docker_enabled: True
      docker_net: bridge
      docker_set_user: 
      docker_sudo: False
      outputs_to_working_directory: False
      require_container: True
      submit_requirements: GalaxyDockerHack == True
      tmp_dir: True
    interactive_pulsar:
      runner: pulsar_embedded
      container_monitor_result: callback
      docker_auto_rm: True
      docker_enabled: True
      docker_net: bridge
      docker_set_user: 
      docker_sudo: False
      outputs_to_working_directory: False
      require_container: True
      submit_requirements: GalaxyDockerHack == True && GalaxyGroup == "interactive"
      tmp_dir: True
    local:
      runner: local
    pulsar_be_tpv:
      runner: pulsar_eu_be01
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    pulsar_it_tpv:
      runner: pulsar_eu_it01
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    pulsar_mira_tpv:
      runner: pulsar_mira_runner
      env:
        - name: GALAXY_MEMORY_MB
          value: {MEMORY_MB}
        - name: GALAXY_SLOTS
          value: {PARALLELISATION}
        - name: LC_ALL
          value: C
        - name: SINGULARITY_CACHEDIR
          value: /data/share/var/database/container_cache
      default_file_action: remote_transfer
      dependency_resolution: none
      jobs_directory: /data/share/staging
      outputs_to_working_directory: false
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      singularity_default_container_id: /cvmfs/singularity.galaxyproject.org/u/b/ubuntu:18.04
      singularity_enabled: True
      submit_request_cpus: {PARALLELISATION}
      submit_request_memory: {MEMORY}
      transport: curl
    remote_cluster_mq_be01:
      runner: pulsar_eu_be01
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    remote_cluster_mq_cz:
      runner: pulsar_eu_cz
      jobs_directory: /storage/brno11-elixir/home/galaxyeu/galaxy-staging/
    remote_cluster_mq_de01:
      runner: pulsar_eu_de01
      env:
        - name: LC_ALL
          value: C
        - name: SINGULARITY_CACHEDIR
          value: /data/share/var/database/container_cache
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      singularity_enabled: true
      singularity_volumes: $defaults
      transport: curl
    remote_cluster_mq_de02:
      runner: pulsar_eu_de02
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    remote_cluster_mq_de03:
      runner: pulsar_eu_de03
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    remote_cluster_mq_docker_de03:
      runner: pulsar_eu_de03
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      docker_auto_rm: true
      docker_default_container_id: ubuntu:18.04
      docker_enabled: true
      docker_run_extra_arguments: --gpus all
      docker_set_user: 
      docker_sudo: false
      docker_volumes: $galaxy_root:ro,$tool_directory:ro,$working_directory:rw,$job_directory:rw,/data/share/staging/:ro
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    remote_cluster_mq_fr01:
      runner: pulsar_eu_fr01
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    remote_cluster_mq_it01:
      runner: pulsar_eu_it01
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    remote_cluster_mq_test:
      runner: pulsar_eu_test
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    remote_cluster_mq_uk01:
      runner: pulsar_eu_uk01
      env:
        - name: LC_ALL
          value: C
      default_file_action: remote_transfer
      dependency_resolution: remote
      jobs_directory: /data/share/staging
      persistence_directory: /data/share/persisted_data
      remote_metadata: false
      rewrite_parameters: true
      transport: curl
    tpv_dispatcher:
      runner: dynamic
      type: python
      function: map_tool_to_destination
      rules_module: tpv.rules
      tpv_config_files:
        - /home/jose/Dokumente/Repositories/usegalaxy-eu/infrastructure-playbook/files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/tool_defaults.yml
        - https://raw.githubusercontent.com/galaxyproject/tpv-shared-database/main/tools.yml
        - /home/jose/Dokumente/Repositories/usegalaxy-eu/infrastructure-playbook/files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/destinations.yml
        - /home/jose/Dokumente/Repositories/usegalaxy-eu/infrastructure-playbook/files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/tools.yml
        - /home/jose/Dokumente/Repositories/usegalaxy-eu/infrastructure-playbook/files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/interactive_tools.yml
        - /home/jose/Dokumente/Repositories/usegalaxy-eu/infrastructure-playbook/files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/users.yml
        - /home/jose/Dokumente/Repositories/usegalaxy-eu/infrastructure-playbook/files/galaxy/dynamic_rules/usegalaxy/total_perspective_vortex/roles.yml



#
# Job mapping configuration
#
tools: []

#
# Job limits configuration
#
limits:
  - type: anonymous_user_concurrent_jobs
    value: 3
  - type: destination_user_concurrent_jobs
    value: 10
    id: condor_upload
  - type: destination_user_concurrent_jobs
    value: 4
    id: interactive_pulsar
  - type: destination_user_concurrent_jobs
    value: 4
    id: interactive_pulsar_hi_mem
  - type: destination_user_concurrent_jobs
    value: 3
    id: sh_fdid_condor_docker_ie
  - type: destination_user_concurrent_jobs
    value: 3
    id: sh_fdid_condor_docker_ie_interactive
  - type: destination_user_concurrent_jobs
    value: 10
    id: sh_fdid_condor_intensive_jobs
  - type: output_size
    value: '300GB'
  - type: registered_user_concurrent_jobs
    value: 30
  - type: walltime
    value: '148:00:00'
```

mira-miracoli commented on group_vars/sn07.yml:
```diff
@@ -431,6 +431,8 @@ galaxy_config_files:
   - src: "{{ galaxy_config_file_src_dir }}/config/echo_main_env.xml"
     dest: "{{ galaxy_config_dir }}/echo_main_env.xml"
+  - src: "{{ galaxy_config_file_src_dir }}/dynamic_rules/usegalaxy/total_perspective_vortex/tool_defaults.yml"
+    dest: "{{ tpv_mutable_dir }}/tool_defaults.yml"    
```
```suggestion
    dest: "{{ tpv_mutable_dir }}/tool_defaults.yml"
```

kysrpex replied:
This should have been fixed in [9790734](https://github.com/usegalaxy-eu/infrastructure-playbook/pull/689/commits/979073442aa293068a521fcc02e986aa5b88eaef) already.
